### PR TITLE
Refactor InitCodeTransformer with context pattern.

### PR DESCRIPTION
**Pull request by shinfan**
- Provide single entry point of init code generation via `InitCodeContext`
- Unified `InitTreeParserContext` with `InitCodeContext`

### CODE REVIEW

File: src/main/java/com/google/api/codegen/transformer/InitCodeTransformer.java [hunk 1]
```diff
@@ -51,106 +51,34 @@
...
-                .build());
-    return buildInitCodeViewFlattened(context, rootNode);
+  private static InitCodeNode createRootNode(InitCodeTransformerContext context) {
+    MethodTransformerContext methodContext = context.methodContext();
+    return InitCodeNode.createTree(
+        InitTreeParserContext.newBuilder()
+            .table(context.symbolTable())
+            .rootObjectType(context.initObjectType())
+            .initValueConfigMap(createCollectionMap(methodContext))
+            .initFieldConfigStrings(methodContext.getMethodConfig().getSampleCodeInitFields())
```
Comment (michaelbausor) [inline src/main/java/com/google/api/codegen/transformer/InitCodeTransformer.java]: Should this be context.getSampleCodeInitFields()?

**Comment by shinfan** [comment]
@michaelbausor I think it might be a good idea. I will try to replace InitTreeParserContext with InitCodeTransformerContext and see how it goes.

File: src/main/java/com/google/api/codegen/transformer/java/JavaGapicSurfaceTestTransformer.java [hunk 2]
```diff
@@ -208,27 +223,25 @@ private GapicSurfaceTestCaseView createTestCaseView(
...
-          initCodeTransformer.generateRequestObjectTestInitCode(
-              methodContext, initSymbolTable, valueGenerator);
+          InitCodeTransformer.generateInitCode(
+              createRequestInitCodeContext(methodContext, initSymbolTable, null, false));
     } else {
       initCodeView =
-          initCodeTransformer.generateFlatteningTestInitCode(
-              methodContext, paramFields, initSymbolTable, valueGenerator);
+          InitCodeTransformer.generateInitCode(
+              createRequestInitCodeContext(methodContext, initSymbolTable, paramFields, true));
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/transformer/java/JavaGapicSurfaceTestTransformer.java]: I can't tell what true/false means here. I would recommend an enum named something like "Flattening".

File: src/main/java/com/google/api/codegen/transformer/InitCodeTransformer.java [hunk 3]
```diff
@@ -51,106 +51,34 @@
...
-
-  public InitCodeView generateMockResponseObjectInitCode(
-      MethodTransformerContext context, SymbolTable table, TestValueGenerator valueGenerator) {
-    ArrayList<Field> primitiveFields = new ArrayList<>();
-    for (Field field : context.getMethod().getOutputMessage().getFields()) {
-      if (field.getType().isPrimitive() && !field.getType().isRepeated()) {
-        primitiveFields.add(field);
-      }
+  /** Generates initialization code from the given InitCodeTransformerContext object. */
+  public static InitCodeView generateInitCode(InitCodeTransformerContext context) {
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/transformer/InitCodeTransformer.java]: Why are you switching to static functions? All of the other transformers use instance methods.

File: src/main/java/com/google/api/codegen/transformer/InitCodeTransformerContext.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+  public abstract boolean isFlattened();
+
+  @Nullable
+  public abstract TestValueGenerator valueGenerator();
+
+  @Nullable
+  public abstract Iterable<String> initFieldConfigStrings();
+
+  @Nullable
+  public abstract Iterable<InitCodeNode> additionalNodes();
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/transformer/InitCodeTransformerContext.java]: make name more specific: `additionalInitCodeNodes`

File: src/main/java/com/google/api/codegen/transformer/InitCodeTransformer.java [hunk 5]
```diff
@@ -51,106 +51,34 @@
...
+        InitTreeParserContext.newBuilder()
+            .table(context.symbolTable())
+            .rootObjectType(context.initObjectType())
+            .initValueConfigMap(createCollectionMap(methodContext))
+            .initFieldConfigStrings(methodContext.getMethodConfig().getSampleCodeInitFields())
+            .initFields(context.fields())
+            .suggestedName(context.suggestedName())
+            .valueGenerator(context.valueGenerator())
+            .additionalSubTrees(context.additionalNodes())
+            .build());
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/transformer/InitCodeTransformer.java]: The InitCodeTransformerContext looks very similar to InitTreeParserContext - do we really need both?

File: src/main/java/com/google/api/codegen/transformer/InitCodeTransformer.java (hunk 3 above)
Comment (shinfan) [inline src/main/java/com/google/api/codegen/transformer/InitCodeTransformer.java]: I switched these methods to static so the caller does not need to initialize an instance of InitCodeTransformer.

File: src/main/java/com/google/api/codegen/transformer/InitCodeTransformerContext.java (hunk 4 above)
Comment (shinfan) [inline src/main/java/com/google/api/codegen/transformer/InitCodeTransformerContext.java]: Done

File: src/main/java/com/google/api/codegen/transformer/java/JavaGapicSurfaceTestTransformer.java (hunk 2 above)
Comment (shinfan) [inline src/main/java/com/google/api/codegen/transformer/java/JavaGapicSurfaceTestTransformer.java]: Done.

**Comment by shinfan** [comment]
@michaelbausor @garrettjonesgoogle  PTAL

File: src/main/java/com/google/api/codegen/DocConfig.java [hunk 6]
```diff
@@ -108,14 +106,14 @@ public BuilderType setFieldInitCode(
...
           InitCodeNode.createTree(
-              InitTreeParserContext.newBuilder()
-                  .table(new SymbolTable())
-                  .rootObjectType(method.getInputType())
-                  .initValueConfigMap(initValueConfigMap.build())
+              InitCodeContext.newBuilder()
+                  .initObjectType(method.getInputType())
                   .initFieldConfigStrings(methodConfig.getSampleCodeInitFields())
-                  .initFields(fields)
+                  .fields(fields)
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/DocConfig.java]: Why rename `initFields` to `fields`?

File: src/main/java/com/google/api/codegen/metacode/InitCodeContext.java [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+import com.google.api.codegen.config.CollectionConfig;
+import com.google.api.codegen.config.InterfaceConfig;
+import com.google.api.codegen.config.MethodConfig;
+import com.google.api.codegen.metacode.InitCodeNode;
+import com.google.api.codegen.metacode.InitValueConfig;
+import com.google.api.codegen.transformer.MethodTransformerContext;
+import com.google.api.codegen.util.Name;
+import com.google.api.codegen.util.SymbolTable;
+import com.google.api.codegen.util.testing.TestValueGenerator;
+import com.google.api.tools.framework.model.*;
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/metacode/InitCodeContext.java]: Don't use * imports

File: src/main/java/com/google/api/codegen/metacode/InitCodeContext.java [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+package com.google.api.codegen.metacode;
+
+import com.google.api.codegen.InterfaceView;
+import com.google.api.codegen.config.ApiConfig;
+import com.google.api.codegen.config.CollectionConfig;
+import com.google.api.codegen.config.InterfaceConfig;
+import com.google.api.codegen.config.MethodConfig;
+import com.google.api.codegen.metacode.InitCodeNode;
+import com.google.api.codegen.metacode.InitValueConfig;
+import com.google.api.codegen.transformer.MethodTransformerContext;
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/metacode/InitCodeContext.java]: codegen.metacode should not depend on anything in codegen.transformer.

File: src/main/java/com/google/api/codegen/transformer/InitCodeTransformer.java [hunk 9]
```diff
@@ -51,115 +47,28 @@
...
-
-  public InitCodeView generateMockResponseObjectInitCode(
-      MethodTransformerContext context, SymbolTable table, TestValueGenerator valueGenerator) {
-    ArrayList<Field> primitiveFields = new ArrayList<>();
-    for (Field field : context.getMethod().getOutputMessage().getFields()) {
-      if (field.getType().isPrimitive() && !field.getType().isRepeated()) {
-        primitiveFields.add(field);
-      }
+  /** Generates initialization code from the given InitCodeContext object. */
+  public static InitCodeView generateInitCode(InitCodeContext context) {
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/transformer/InitCodeTransformer.java]: I want the classes to work consistently. This would be the only transformer class with static methods. Either they should all have static methods, or they should all have instance methods. I don't think it is something you should change under the scope of this current review.

Comment (shinfan) [inline src/main/java/com/google/api/codegen/transformer/InitCodeTransformer.java]: I guess consistency is more important here. Reverted.

File: src/main/java/com/google/api/codegen/DocConfig.java (hunk 6 above)
Comment (shinfan) [inline src/main/java/com/google/api/codegen/DocConfig.java]: Changed back to initFields. Done.

File: src/main/java/com/google/api/codegen/metacode/InitCodeContext.java [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+import java.util.List;
+import java.util.Map;
+
+/** The context used for generating init code views */
+@AutoValue
+public abstract class InitCodeContext {
+  public enum InitCodeParamType {
+    RequestParam,
+    FlattenedParam,
+  }
```
Comment (shinfan) [inline src/main/java/com/google/api/codegen/metacode/InitCodeContext.java]: Done.

File: src/main/java/com/google/api/codegen/metacode/InitCodeContext.java (hunk 8 above)
Comment (shinfan) [inline src/main/java/com/google/api/codegen/metacode/InitCodeContext.java]: Done. Separate out `MethodTransformerContext` from `InitCodeContext`

File: src/main/java/com/google/api/codegen/metacode/InitCodeContext.java [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+
+  /**
+   * Returns the output type of the init code. Default to SingleObject.
+   */
+  public abstract InitCodeOutputType outputType();
+
+  @Nullable
+  /**
+   *  Returns the value generator which is used to produce deterministically random unique
+   *  values for testing purpose.
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/metacode/InitCodeContext.java]: s/purpose/purposes/

Also, `@Nullable` should go next to the method - the comment shouldn't be in the middle.

File: src/main/java/com/google/api/codegen/metacode/InitCodeContext.java [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+
+  /**
+   * The symbol table used to store unique symbols used in the init code.
+   * Default to empty table.
+   */
+  public abstract SymbolTable symbolTable();
+
+  @Nullable
+  /**
+   * Contains the fields that requires init.
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/metacode/InitCodeContext.java]: s/that requires/that require/

File: src/main/java/com/google/api/codegen/metacode/InitCodeContext.java [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+
+  @Nullable
+  /**
+   *  Returns the value generator which is used to produce deterministically random unique
+   *  values for testing purpose.
+   */
+  public abstract TestValueGenerator valueGenerator();
+
+  @Nullable
+  /** Returns init config strings from user config file.*/
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/metacode/InitCodeContext.java]: Remove the "from user config file part" - that is outside the scope of this class.

File: src/main/java/com/google/api/codegen/transformer/InitCodeTransformer.java [hunk 14]
```diff
@@ -51,115 +47,30 @@
...
-      MethodTransformerContext context, SymbolTable table, TestValueGenerator valueGenerator) {
-    ArrayList<Field> primitiveFields = new ArrayList<>();
-    for (Field field : context.getMethod().getOutputMessage().getFields()) {
-      if (field.getType().isPrimitive() && !field.getType().isRepeated()) {
-        primitiveFields.add(field);
-      }
+  /**
+   * Generates initialization code from the given MethodTransformerContext and InitCodeContext
+   * objects.
+   * */
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/transformer/InitCodeTransformer.java]: another place with a malformed comment (and there are two more)

File: src/main/java/com/google/api/codegen/metacode/InitCodeContext.java (hunk 12 above)
Comment (shinfan) [inline src/main/java/com/google/api/codegen/metacode/InitCodeContext.java]: Done

File: src/main/java/com/google/api/codegen/metacode/InitCodeContext.java (hunk 11 above)
Comment (shinfan) [inline src/main/java/com/google/api/codegen/metacode/InitCodeContext.java]: Done

File: src/main/java/com/google/api/codegen/metacode/InitCodeContext.java (hunk 13 above)
Comment (shinfan) [inline src/main/java/com/google/api/codegen/metacode/InitCodeContext.java]: Done

**Comment by shinfan** [comment]
PTAL

**Comment by garrettjonesgoogle** [comment]
LGTM
